### PR TITLE
Add a deprecated timestamp

### LINKED ISSUE #705: [DISCOVERY] Service Lifecycle Clarity
The service definition does not support the state of a service (ACTIVE, DEPRECATED, RETIRED) etc.

Clients should be able to search/filter by such states enabling them to proactively plan to move off services/endpoints as they are decommissioned. Ideally a state of DEPRECATED would be associated with a date for final retirement.

Additionally (and maybe this is a separate issue).. Is a _subscriptionurl_ immutable? If not what how would a client know that it has changed, and what notice would they get that it's about to change?.

### PR DESCRIPTION
**Pull request by duglin**
Closes #705

Signed-off-by: Doug Davis <dug@us.ibm.com>

### CODE REVIEW

File: discovery.md [hunk 1]
```diff
@@ -136,6 +136,7 @@ Service:
   "url": "[unique URL to this service]",
   "description": "[human string]", ?
   "docsurl": "[URL reference for human documentation]", ?
+  "removaltime": "[RFC339 timestamp of when service will be removed]", ?
```
Comment (tomkerkhove) [inline discovery.md:139]: Shouldn't we give guidance on what to migrate to?
```suggestion
  "removal": {
    "time": "[RFC339 timestamp of when service will be removed]", ?
    "alternative": "", ?
  }
```

Comment (duglin) [reply discovery.md:139]: Questions from the group:

John:
- what kind of guarantees are there around "alt"?  does it have it to exist before the time?
- do subscriptions carry over to the alt?

- Erik: any kind of transition?

Jennifer:
- what's the value? id or url? machine or human readable?
- or how does someone interpret it?

Comment (duglin) [reply discovery.md:139]: My recommendation, if we choose to include "alt":
- the "alt" must be available immediately 
- subscriptions do not carry over and there's no guarantee that the 'alt' is a semantic replacement for the service. It is simply a recommendation for consideration and the user is expected to investigate if it's a good match.
- no transitioning or transferring of subscriptions to "alt"
- "alt" must be the URL of the alt service. MAY be in another DE, but no guarantee that the client can access it.

Comment (duglin) [reply discovery.md:139]: s/removal/deprecated/

**Comment by duglin** [comment]
ok ready for another round of reviews
@cloudevents/notify

File: discovery/spec.md [hunk 2]
```diff
@@ -132,6 +132,10 @@ Service:
   "url": "[unique URL to this service]",
   "description": "[human string]", ?
   "docsurl": "[URL reference for human documentation]", ?
+  "deprecated": { ?
+    "removaltime": "[RFC339 timestamp of when service will be removed]", ?
```
Comment (JemDay) [inline discovery/spec.md:136]: Following the spirit of the main comment thread to follow the HTTP semantics shouldn't we call this `sunsettime` ??

**Review by JemDay — COMMENTED** [review]
Suggest following the other convention mention to use the term `sunset` to represent the point at which the item is no longer available.

File: discovery/spec.md (hunk 2 above)
Comment (duglin) [reply discovery/spec.md]: done

**Comment by duglin** [comment]
This one is ready for review on this week's call.
@cloudevents/notify

File: discovery/discovery.yaml [hunk 3]
```diff
@@ -238,6 +238,15 @@ components:
           type: string
           description: Absolute URL that provides a link to additional documentation about the service. This is intended for a developer to use in order to learn more about this service's events produced
           example: http://cloud.example.com/docs/blobstorage
+        deprecated:
+          type: object
+          properties:
+            sunsettime:
+              type: string
+              description: The time at which the service will be removed
+            alterntaive:
```
Comment (jskeet) [inline discovery/discovery.yaml:247]: Typo

File: discovery/spec.md [hunk 4]
```diff
@@ -304,6 +308,44 @@ The following sections define the attributes that appear in a Service entity.
 - Examples:
   - `http://cloud.example.com/docs/blobstorage`
 
+##### deprecated
+
+- Type: Object containing 2 properties:
+  - sunsettime<br>
+    An OPTIONAL property indicating the time when the Service MAY be removed.
```
Comment (jskeet) [inline discovery/spec.md:315]: Given that we don't use the word "sunset" anywhere in the description, maybe "removaltime"? That also avoids the easily-typo'd double t.

File: discovery/spec.md [hunk 5]
```diff
@@ -304,6 +308,44 @@ The following sections define the attributes that appear in a Service entity.
...
+    The Service MUST NOT be deleted before this time. If this property is not
+    present then client can not make any assumption as to when the Service
+    might be removed. Note: as with most Service properties, this property
+    is mutable. If present, this MUST be an [RFC339][rfc3339] timestamp.
```
Comment (duglin) [inline discovery/spec.md:319]: Add a 'started' time

File: discovery/spec.md (hunk 4 above)
Comment (jskeet) [reply discovery/spec.md:315]: Possibly earliestremovaltime to indicate that it's a lower bound...

File: discovery/discovery.yaml (hunk 3 above)
Comment (duglin) [inline discovery/discovery.yaml:247]: typo

File: discovery/spec.md (hunk 2 above)
Comment (duglin) [reply discovery/spec.md:137]: per the call today we switched to "removaltime"

**Comment by duglin** [comment]
Made all of the updates based on today's call.

**Comment by duglin** [comment]
approved on the 5/5 call

**Pull request merged**